### ISSUE TITLE
[Feature Request] Support matching by user name (not just user ID) and/or cgroup name

### ISSUE THREAD

**Issue by quentinmit** [evilsocket/opensnitch#1116]
### Summary:
`systemd` often uses dynamically-chosen UIDs for services ([`DynamicUser=yes`](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#User/Group%20Identity)). This means that rules matching with the `uid` operand will no longer match when a service/the system is next restarted and the process gets a newly-assigned uid. As a consequence, I have to allow a binary for all users on the system, instead of just allowing a daemon user to make connections with a binary.

At process discovery, OpenSnitch should use `user.LookupId(uid)` to look up the name currently assigned to a uid, and allow the result to be matched with an operand.

In addition, or perhaps in lieu, it would be helpful to match by a process's cgroup name. systemd will assign a unique cgroup to each service, so this would ensure that even if a process happens to somehow be launched as the correct user, it will still not be able to evade rules. This information is available in `/proc/$pid/cgroup`. For example:

```
$ cat /proc/$(systemctl show -P ExecMainPID sshd)/cgroup
0::/system.slice/sshd.service
$ cat /proc/self/cgroup 
0::/user.slice/user-1000.slice/user@1000.service/app.slice/app-konsole-3d42e27e456a4ce9a1ad5982ed8f6a33.scope
```

This could be especially useful for desktop apps, because you could allow `wget` to be used from Konsole without also allowing it to be used by Spotify.

**Comment by gustavo-iniguez-goya** [comment]
I've added option to filter by user name.

However in the scenario described, using DynamicUser=, it won't work. On the one hand because usernames are allocated dynamically and doesn't exist in the system, so user.LookupId() won't find it when loading the rule.

On the other hand, if User= is specified and it doesn't exist in the system, user.LookupId() won't find the UID for the same reason.

https://cs.opensource.google/go/go/+/master:src/os/user/lookup_unix.go;l=218?q=lookupUser&sq=&ss=go%2Fgo
https://cs.opensource.google/go/go/+/master:src/os/user/lookup.go;l=10?q=userFile&ss=go%2Fgo

**Comment by quentinmit** [comment]
> However in the scenario described, using DynamicUser=, it won't work. On the one hand because usernames are allocated dynamically and doesn't exist in the system, so user.LookupId() won't find it when loading the rule.

You need to perform the lookup at rule evaluation time, not when loading the rule (because the user won't exist until the service starts up).

> On the other hand, if User= is specified and it doesn't exist in the system, user.LookupId() won't find the UID for the same reason.
> 
> https://cs.opensource.google/go/go/+/master:src/os/user/lookup_unix.go;l=218?q=lookupUser&sq=&ss=go%2Fgo https://cs.opensource.google/go/go/+/master:src/os/user/lookup.go;l=10?q=userFile&ss=go%2Fgo

The implementation you linked is only used when the binary is built without Cgo support. When built with Cgo, [this implementation](https://cs.opensource.google/go/go/+/master:src/os/user/cgo_lookup_unix.go;l=22) is used instead, which should correctly resolve usernames via libc and [nss-systemd](https://www.freedesktop.org/software/systemd/man/latest/nss-systemd.html).

**Comment by gustavo-iniguez-goya** [comment]
Thanks @quentinmit . In general we try to "compile" the rules to avoid having to make excessive calls when evaluating them (for example by calling `user.Lookup()` only when loading the rule to compare only the UID per iteration/connection, instead of calling `user.Lookup()` per connection).

In this case, calling `user.LookupId()` per connection, adds an overhead of 40-150us per connection and rule in my system (a rule may be evaluated multiple times).

It's something to take into account, but I think it's acceptable if it solves this use case. I'll test it in more detail.

**Comment by gustavo-iniguez-goya** [comment]
@quentinmit would you mind explaining this use case?

> This could be especially useful for desktop apps, because you could allow wget to be used from Konsole without also allowing it to be used by Spotify.

How do you have configured Konsole/Spotify to distinguish which one is launching wget/curl, etc? Just to reproduce it and test this more properly.

user.LookupId() is working fine, and since these virtual users IDs are always in the range 61184–65519 , we could exclude UIDs out of this range, for some performance gain.

**Comment by CertainLach** [comment]
> How do you have configured Konsole/Spotify to distinguish which one is launching wget/curl, etc? Just to reproduce it and test this more properly.

KDE Plasma, for example, launches every desktop program (by .desktop file) in its own cgroup

> In this case, calling user.LookupId() per connection, adds an overhead of 40-150us per connection and rule in my system (a rule may be evaluated multiple times).

Can't opensnitch daemon just compute and store cgroup name in bpfmap? (This is much harder with users, due to setuid and other calls, but at least it should be possible to cache username in bpfmap, assuming no daemon process will access network before calling setuid)

**Comment by gustavo-iniguez-goya** [comment]
thank you @CertainLach . Probably I didn't pay attention how KDE reports users, but I haven't seen DynamicUsers. I'll double check it 👍 

Regarding to obtain this info from ebpf, I guess so it's possible. I haven't digged into cgroups though.

**Comment by gustavo-iniguez-goya** [comment]
> KDE Plasma, for example, launches every desktop program (by .desktop file) in its own cgroup

At least in KDE 6.2.5, Konsole, Spotify, Firefox or any other apps are launched as the system users, no Dynamic Users , which is what @quentinmit is requesting. 

The apps are certainly launched in different cgroups, but no as dynamic users.

**Comment by gustavo-iniguez-goya** [comment]
yeah, a cache of users would be useful. The only thing to think is when to delete those users from cache.

regarding the cgroups, I think that we could parse /proc/<pid>/cgroup and add a new rule Operand:

```
{
  "type": "simple",
  "operand": "cgroup.path"
  "data": "0::/user.slice/user-1000.slice/user@1000.service/app.slice"
}
```

possible operands: cgroup.path, cgroup.slice, cgroup.scope ...

combined with "type": "regexp" or "type": "list", would allow us to match several components of the cgroup.

It seems easy. We'd obtain the cgroup info from /proc for now, and later from the kernel.

**Comment by quentinmit** [comment]
> regarding the cgroups, I think that we could parse /proc//cgroup and add a new rule Operand:
> 
> ```
> {
>   "type": "simple",
>   "operand": "cgroup.path"
>   "data": "0::/user.slice/user-1000.slice/user@1000.service/app.slice"
> }
> ```
> 
> possible operands: cgroup.path, cgroup.slice, cgroup.scope ...

The /proc file contains multiple lines with 3 separate colon-separated fields (see [`man cgroups`](https://man7.org/linux/man-pages/man7/cgroups.7.html))) - the cgroup ID, controller types, and path. On modern kernels with cgroups v2 (systemd 256 [drops support](https://github.com/systemd/systemd/issues/30852) for cgroups v1), there is only one line, the ID is always 0, and the types are always empty, so really only the path matters. I would parse the path out and use that as the string to match against.

> combined with "type": "regexp" or "type": "list", would allow us to match several components of the cgroup.

IMO a glob-style match would be the best, although that can certainly be emulated with regexp. E.g. `/user.slice/user-*.slice/user@*.service/app.slice/app-org.kde.konsole@**`. 

If you want to split the path into separate components, I suggest using names inspred by the journald field names for the parts (see [`man systemd.journal-fields`](https://man7.org/linux/man-pages/man7/systemd.journal-fields.7.html)) - they call them

```
_SYSTEMD_CGROUP
_SYSTEMD_SLICE
_SYSTEMD_UNIT
_SYSTEMD_USER_UNIT
_SYSTEMD_USER_SLICE
```

but I think it's fine, especially for a first iteration, to only provide matching on the entire path and let the user write a regex/glob that matches as they wish.

> It seems easy. We'd obtain the cgroup info from /proc for now, and later from the kernel.

That would be awesome!

**Comment by LordGrimmauld** [comment]
While testing this, i came across this comment:
https://github.com/evilsocket/opensnitch/blob/7c438de082fc2b3f4272439ed97babe0e9a58727/daemon/rule/operator.go#L121-L125

I have the following rule:
```json
{
  "action": "allow",
  "created": "1970-01-01T00:00:00.0+00:00",
  "duration": "always",
  "enabled": true,
  "name": "nixbld",
  "operator": {
    "data": "nixbld[0-9]+",
    "operand": "user.name",
    "sensitive": true,
    "type": "regexp"
  }
}
```
Motivation here is to allow all `nixbld*` unix users to access the network. The nix evaluator does cursed stuff and allocates a number of distinct users to do sandboxed builds. I wanted to allow all of them network access, but then found out regexp is not yet supported for this application. Is regexp user matching still planned?
